### ISSUE TITLE
Class name stuck to first Stitches-styled component in composition

### ISSUE THREAD

**Issue by LucasUnplugged** [modulz/stitches#675]
# Bug report

## Describe the bug
- If we pass a Stitches-`styled` component as the 1st param to another `styled` expression, the `className` of the parent is used on the child.
- In cases where there are 3+ levels of composition, the `className` of the first component in the chain is used for _all_ composed components.

## To Reproduce

Example: in this [code sandbox](https://codesandbox.io/s/staging-field-4cs36?file=/src/App.tsx), I created a `Box` primitive, and used it for composition in 2 components: `Header` and `Footer`.

I then placed them both inside another component (`Main`), and styled it as such:
```TSX
const Main = styled(Box, {
  background: '$gray500',
  [`& ${Header}`]: {
    background: '$purple600',
    color: 'white',
  },
});
```

## Expected behavior

In that example, only `Header` should be purple with white text. Instead, both `Header` and `Footer` have that styling.

## Screenshots

<img width="900" alt="Screen Shot 2021-07-24 at 8 41 37 PM" src="https://user-images.githubusercontent.com/2473823/126884223-53119b57-b6ab-48b4-ab3e-31b3dfebcd96.png">

## System information

- OS: macOS
- Browser: Chrome 91.0.4472.114 (Official Build) (x86_64)
- Version of Stitches: 0.2.3
- Version of Node.js: 10.20.1, [according to this](https://codesandbox.io/docs/faq#can-i-change-the-node-version-used-in-a-container-sandbox).

## Additional context

The console logs show that the `Box` primitive's `className` is being used on all composed components (see screenshot).

**Comment by rafgraph** [comment]
There's an atomic-like CSS specificity issue (using style objects instead of single rules) that I think is related. It seems like class names are generated based on deep equal (or style equal) for the style object (instead of referential equality), so if there's an object with matching styles someplace else in the app there will be specificity issues.

See below codesandbox for an example. The `div`s and `span`s are completely independent, but because they have deep equal style objects the `span`s render incorrectly.
https://codesandbox.io/s/pensive-microservice-tdb4y?file=/src/App.tsx

**Comment by peduarte** [comment]
@rafgraph Hey, thanks for chiming in. It is a somewhat related issue - but it's different. Applying Lucas' solution won't solve the problem you're demonstrating. [Sandbox using Lucas' fix as dependency](https://codesandbox.io/s/boring-germain-308mj?file=/src/App.tsx)

The underlying cause of both issues is the fact that the hashes are based on the style object. So two totally different elements with the same styles, will get the same class. 

The good thing about this is that overall it generates less CSS, but it also seems quite fragile and can cause unexpected side effects. 

I think it's worth opening a new issue with your demo, so we can track it separately. 

Thank you!

**Comment by rafgraph** [comment]
Sounds good, I opened #690. Yes, this does seem quite brittle and I don't think the tradeoff is worth it. It's kind of back to the atomic CSS issues that early versions Stitches had.

**Comment by LucasUnplugged** [comment]
> See below codesandbox for an example. The `div`s and `span`s are completely independent, but because they have deep equal style objects the `span`s render incorrectly.
> https://codesandbox.io/s/pensive-microservice-tdb4y?file=/src/App.tsx

I think the issue here is sequencing of styles. After doing some testing, I've now learned that Stitches seems to define the sequence of styles based NOT on order of definition, but on **order of usaged** in your markup. See [this fork off @rafgraph's sandbox](https://codesandbox.io/s/crazy-water-gg64r?file=/src/App.tsx). Because I moved `Span1` before `Div1` in the markup, the issue is fixed.

But here again, I wonder how likely this is in real-world use cases; the issue would probably have to happen in the same file/module, and the likelihood of having the exact same styles in two different components, in the same module, seems fairly low to me — you could probably use variants for real-world cases along the lines of that example, which I think would avoid the issue.

**Comment by rafgraph** [comment]
> I think the issue here is sequencing of styles. After doing some testing, I've now learned that Stitches seems to define the sequence of styles based NOT on order of definition, but on order of usaged in your markup. See this fork off @rafgraph's sandbox. Because I moved Span1 before Div1 in the markup, the issue is fixed.

Yes it seems to be based on order of usage. In your solution you have to change the markup which changes what the user sees, so to me this doesn't seem like a good work around.

> the issue would probably have to happen in the same file/module

I don't think this is correct. Here's a [codesandbox with the `div`s and `span`s in separate components in separate files](https://codesandbox.io/s/relaxed-lehmann-kz34s?file=/src/App.tsx), and the issue persists. It seems like any deep equal style objects anywhere in the app will have the same class name. So the larger the app the more likely this bug will be run into (and the more brittle Stitches becomes).

Also, the bug is path dependent. The actions the user takes (the order of things they click on, pages they visit, etc) can influence when this bug pops up (because the styles are added based on render order). See [this codesandbox for an example](https://codesandbox.io/s/agitated-worker-lj8x5?file=/src/App.tsx), if you click "show divs" and then click "show spans" the spans render incorrectly. But if you click "show spans" and then click "show divs" the divs render incorrectly (you'll need to refresh the page between tests).

In my experience path dependent bugs are some of the most confusing and hardest to fix, and preventing them should be given a heavy weighting when considering tradeoffs.

**Comment by LucasUnplugged** [comment]
Sorry, I just started a new job, so I'm only getting back to this now.

TL;DR: full response below, but I agree that #690 is a very serious issue. That said, I would still push for this fix to go in as well, as this one basically prevents composition of Stitches primitives.

> > Because I moved Span1 before Div1 in the markup, the issue is fixed.
> 
> Yes it seems to be based on order of usage. In your solution you have to change the markup which changes what the user sees, so to me this doesn't seem like a good work around.

Oh, I didn't mean the above as a solution to the problem (which I agree, would NOT be a good solution); I was only trying to understand (and show) how it was caused.

> > the issue would probably have to happen in the same file/module
> 
> I don't think this is correct. Here's a [codesandbox with the `div`s and `span`s in separate components in separate files](https://codesandbox.io/s/relaxed-lehmann-kz34s?file=/src/App.tsx), and the issue persists. It seems like any deep equal style objects anywhere in the app will have the same class name. So the larger the app the more likely this bug will be run into (and the more brittle Stitches becomes).

Oh, interesting; it can happen even with just composition, without even using these (styled) components as children in the style object.

> In my experience path dependent bugs are some of the most confusing and hardest to fix, and preventing them should be given a heavy weighting when considering tradeoffs.

Yes, given my experiment above (i.e., seeing that the order of the markup matters for this bug), I'm not surprised that it is path dependent as well. It looks like Stitches builds up the pool of class names as components (functions) are called, which is smart/sensible. And I agree that'd be a nightmare to debug.

The entire issue you found, however, is separate from this one, as @peduarte and you have said — though I agree that it is perhaps even more serious. It might be trickier to solve, though; I think it comes down to needing some way to make the class names more unique, but I'll comment on your issue about that.

**Comment by LucasUnplugged** [comment]
This issue was fixed by the v1 release! 🎉

[Sample sandbox](https://codesandbox.io/s/stitches-ci-cra-forked-zwdqr?file=/src/App.tsx) showing it working just fine.
